stdpopsim/ext/selection.py

```diff
@@ -151,3 +190,195 @@ def __attrs_post_init__(self):
+        zero and one. Defaults to `0.5`.
+    :param float start_time: The number of generations in the past at which the
+        sweep starts and the mutation switches from neutral to beneficial. Defaults to
+        `mutation_time` (e.g. a "hard" sweep).
+    :param float end_time: The number of generations in the past at which the
+        sweep ends and the mutation switches from beneficial to neutral. Defaults to
+        `0` (e.g. the end of the simulation).
+    :param float min_freq_at_start: The minimum allowed frequency for the
+        mutation in population `population_id` at the start of the sweep. Defaults
+        to `0.0`.
+    :param float min_freq_at_end: The minimum allowed frequency for the
+        mutation in population `population_id` at the end of the sweep. Defaults to
+        `0.0`.
+    """
+
+    # TODO: What about metadata? Currently single sites are stored as strictly
+    # neutral DFEs, and no information is recorded regarding SLiM events that would
+    # change the fitness of the drawn mutation.
+
+    if start_time is None:
+        start_time = mutation_time
+    if end_time is None:
+        end_time = 0
+
+    validate_time_range(mutation_time, start_time)
+    validate_time_range(start_time, end_time)
+    if selection_coeff < 0.0:
+        raise ValueError("Selection coefficient must be non-negative.")
+
+    slim_events = [
+        DrawMutation(
+            single_site_id=single_site_id,
+            time=mutation_time,
+            population_id=population_id,
+        ),
+        ConditionOnAlleleFrequency(
+            single_site_id=single_site_id,
+            start_time=GenerationAfter(mutation_time),
+            end_time=end_time,
+            population_id=population_id,
+            allele_frequency=0.0,
+            op=">",
+        ),
+    ]
+
+    # Neutral period between mutation introduction and onset of sweep.
+    if mutation_time > start_time:
+        slim_events.extend(
+            [
+                ChangeMutationFitness(
+                    single_site_id=single_site_id,
+                    start_time=mutation_time,
+                    end_time=start_time,
+                    population_id=population_id,
+                    selection_coeff=0.0,
+                    dominance_coeff=dominance_coeff,
+                ),
+            ]
+        )
+
+    # Sweep period. Only check sweep AF conditions in start/end generations, so
+    # that allele trajectories are otherwise unconstrained.
+    slim_events.extend(
+        [
+            ChangeMutationFitness(
+                single_site_id=single_site_id,
+                start_time=start_time,
+                end_time=end_time,
+                population_id=population_id,
+                selection_coeff=selection_coeff,
+                dominance_coeff=dominance_coeff,
+            ),
+        ]
+    )
+
+    if min_freq_at_start is not None:
+        if not 0 < min_freq_at_start <= 1:
```

Comment:
it should be allowed to set `min_freq_at_start = 0.0`? I'd do this like
```
if min_freq_at_start is None:
    min_freq_at_start = 0.0
if min_freq_at_start > 0:
   # ...
```

Reply:
I agree @mufernando, `min_freq_at_XXX` implies a closed interval and as the mutation can never be lost, it's gotta be >0. I'll change the docs instead.